src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.RegularExpressions
+{
+    internal enum RegexParseError
+    {
+        TooManyAlternates,
+        IllegalCondition,
+        IncompleteSlashP,
+        MalformedSlashP,
+        UnrecognizedEscape,
+        UnrecognizedControl,
+        MissingControl,
+        TooFewHex,
+        CaptureGroupOutOfRange,
+        UndefinedNameRef,
+        UndefinedBackref,
+        MalformedNameRef,
+        IllegalEndEscape,
+        UnterminatedComment,
+        UnrecognizedGrouping,
+        AlternationCantCapture,
+        AlternationCantHaveComment,
+        MalformedReference,
+        UndefinedReference,
+        InvalidGroupName,
+        CapnumNotZero,
+        UnterminatedBracket,
+        SubtractionMustBeLast,
+        ReversedCharRange,
+        BadClassInCharRange,
+        NotEnoughParens,
+        IllegalRange,
+        InternalError,
+        NestedQuantify,
+        QuantifyAfterNothing,
+        TooManyParens,
+        UnknownProperty // Unicode block, \p{Property}
+    }
+
+    internal class RegexParseException : ArgumentException
```

Comment:
The owners of the System.Xml parser did not want to include error codes for exactly this reason. Given we're now SxS, the concern about breaks is a little lower. Certainly changing the error code in an exception is much less likely to break folks than changing a success case to an error case. And changing to another existing code a little less likely than changing to a newly added error code. I think it's probably OK.

Reply:
Note: it's potentially worth pointing out that Roslyn (which exposes parses and handles tons of input error cases) *specifically* does not expose any type of enum for this sort of thing.  Turns out, in practice, it's not that useful, and it's a bit of a maintenance burden.  

instead, it simply returns back trees along with "diagnostics" which are effectively opaque objects that just say "here's a message about what's wrong, and here's the location it happened".  That's been highly sufficient for our needs (and we specifically expose this stuff so people can analyze code effectively).

This is also the model i took for the new Roslyn regex parser.  

1. outside of stack-overflows, you *always* get a tree back.
2. the tree always contains all input text of the original user input (including whitespace/comments).
3. issues are returned as diagnostics that are provided along with the tree.  diagnostics are simply a message, along with a position/span of relevance.  This is used for doing things like placing a squiggle appropriately.

Reply:
> How did you end up doing this -- did you copy and instrument our parser, or write your own?

i rewrote it entirely, using the reference specification, and then diving into your impl to see how things deviated at times.  For example, character class parsing is exceptionally strange in the native parser.  But i replicated it in on our end because compatibility with the core parser is the #1 priority.  Tons of tests were added (i.e. the basis of htis PR) to validate the same observable behavior between the two systems.

Note: i originally tried to just fork and use the native code.  However, it was far too dfificult as the existing code is designed to generate a lossy AST in place, which it continually mutates as it proceeds.  Trying to work off of that was just too difficult.  

When necessary, i also took code wholesale.  For example, the unicode handling code was copied practically verbatim here: https://github.com/dotnet/roslyn/pull/23984/files#diff-f689c52f48c471beeaea4252031ed8f3

Reply:
> It seems like you're signign yourself up to very specific ways of parsing in order to preserve the error codes you emit.

Given it's a field on an internal type, I do think we would be OK with letting this value change with parser modifications. If the field is public I think it's something to think about.

Reply:
The point why I added the enum in the first place was that I wanted to assert that specific patterns cause specific errors. With that I can be more precise and catch only cases that I expect to fail. I noticed that in your test bed you catched on various error cases like ArgumentException, InvalidOperationException, ... globally.

Reply:
Note: my tests only cared about ArgumentException.  The other exceptions were catching the internal parser errors that i reported and dan previously fixed :)

The enum would certainly be helpful, to ensure that the expected error is happening, and it's not some other 'argument' issue.  However, as this enum is internal, that won't really help out roslyn that much :)   Given that, i have no horse in this race since this choice is entirely internal and just about helping *you* test things.

If you do end up making these public, i would have a lot more feedback on things.

Reply:
Right, the current goal is to enrich the current test suite with your tests. At a later point we can definitely talk about exposing more diagnostics to converge different parser implementations.